File: app/src/main/java/com/quranapp/android/views/homepage/ReadHistoryLayout.java
/*
 * Copyright (c) Faisal Khan (https://github.com/faisalcodes)
 * Created on 28/2/2022.
 * All rights reserved.
 */

package com.quranapp.android.views.homepage;

import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;
import static android.view.ViewGroup.LayoutParams.WRAP_CONTENT;
import static com.peacedesign.android.utils.Dimen.dp2px;

import android.content.Context;
import android.content.Intent;
import android.graphics.Typeface;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import com.peacedesign.android.utils.ViewUtils;
import com.quranapp.android.R;
import com.quranapp.android.activities.ActivityReadHistory;
import com.quranapp.android.adapters.ADPReadHistory;
import com.quranapp.android.components.quran.QuranMeta;
import com.quranapp.android.components.readHistory.ReadHistoryModel;
import com.quranapp.android.databinding.LytHomepageTitledItemTitleBinding;
import com.quranapp.android.db.readHistory.ReadHistoryDBHelper;
import com.quranapp.android.interfaceUtils.Destroyable;
import com.quranapp.android.utils.extensions.TextViewKt;
import com.quranapp.android.utils.extensions.ViewKt;
import com.quranapp.android.utils.extensions.ViewPaddingKt;
import com.quranapp.android.utils.thread.runner.CallableTaskRunner;
import com.quranapp.android.utils.thread.tasks.BaseCallableTask;

import java.util.List;

public class ReadHistoryLayout extends HomepageCollectionLayoutBase implements Destroyable {
    private final CallableTaskRunner<List<ReadHistoryModel>> taskRunner = new CallableTaskRunner<>();
    private final ReadHistoryDBHelper mReadHistoryDBHelper;
    private ADPReadHistory mAdapter;
    private boolean mFirstTime = true;

    public ReadHistoryLayout(@NonNull Context context) {
        this(context, null);
    }

    public ReadHistoryLayout(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ReadHistoryLayout(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mReadHistoryDBHelper = new ReadHistoryDBHelper(context);
    }

    @Override
    public void destroy() {
        if (mReadHistoryDBHelper != null) {
            mReadHistoryDBHelper.close();
        }
    }

    @Override
    protected int getHeaderTitle() {
        return R.string.strTitleReadHistory;
    }

    @Override
    protected int getHeaderIcon() {
        return R.drawable.dr_icon_history;
    }

    @Override
    protected boolean showViewAllBtn() {
        return true;
    }

    @Override
    protected void setupHeader(Context context, LytHomepageTitledItemTitleBinding header) {
        header.titleIcon.setColorFilter(ContextCompat.getColor(context, R.color.colorPrimary));
    }

    @Override
    protected void onViewAllClick(Context context) {
        context.startActivity(new Intent(context, ActivityReadHistory.class));
    }

    private void refreshHistory(QuranMeta quranMeta) {
        if (!taskRunner.isDone()) {
            taskRunner.cancel();
        }

        taskRunner.callAsync(new BaseCallableTask<List<ReadHistoryModel>>() {
            @Override
            public void preExecute() {
                if (mFirstTime) {
                    showLoader();
                    mFirstTime = false;
                }
            }

            @Override
            public List<ReadHistoryModel> call() {
                return mReadHistoryDBHelper.getAllHistories(10);
            }

            @Override
            public void onComplete(List<ReadHistoryModel> models) {
                hideLoader();

                if (models.size() == 0) {
                    makeNoHistoryAlert();
                } else {
                    if (mAdapter == null) {
                        mAdapter = new ADPReadHistory(getContext(), quranMeta, models, dp2px(getContext(), 280));
                        resolveListView().setAdapter(mAdapter);
                    } else {
                        mAdapter.updateModels(models);
                    }
                }
            }
        });
    }

    private void makeNoHistoryAlert() {
        View list = findViewById(R.id.list);
        if (list != null) {
            ViewKt.removeView(list);
        }

        if (findViewById(R.id.text) != null) {
            return;
        }

        Context ctx = getContext();
        AppCompatTextView tv = new AppCompatTextView(ctx);
        tv.setId(R.id.text);

        ViewPaddingKt.updatePaddings(tv, dp2px(ctx, 20), dp2px(ctx, 15));
        TextViewKt.setTextSizePx(tv, R.dimen.dmnCommonSize1_5);
        TextViewKt.setTextColorResource(tv, R.color.colorText2);

        tv.setTypeface(Typeface.SANS_SERIF, Typeface.ITALIC);
        tv.setLayoutParams(new ViewGroup.LayoutParams(MATCH_PARENT, WRAP_CONTENT));
        tv.setText(R.string.strMsgReadShowupHere);
        addView(tv);
    }

    @Override
    protected RecyclerView resolveListView() {
        ViewKt.removeView(findViewById(R.id.text));
        return super.resolveListView();
    }

    public void refresh(QuranMeta quranMeta) {
        refreshHistory(quranMeta);
    }
}

File: app/src/main/java/com/quranapp/android/views/homepage/ReadHistoryLayout.kt
package com.quranapp.android.views.homepage

import android.content.Context
import android.content.Intent
import android.graphics.Typeface
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.LayoutParams.MATCH_PARENT
import android.view.ViewGroup.LayoutParams.WRAP_CONTENT
import androidx.appcompat.widget.AppCompatTextView
import androidx.recyclerview.widget.RecyclerView
import com.peacedesign.android.utils.Dimen
import com.quranapp.android.R
import com.quranapp.android.activities.ActivityReadHistory
import com.quranapp.android.adapters.ADPReadHistory
import com.quranapp.android.components.quran.QuranMeta
import com.quranapp.android.databinding.LytHomepageTitledItemTitleBinding
import com.quranapp.android.db.readHistory.ReadHistoryDBHelper
import com.quranapp.android.interfaceUtils.Destroyable
import com.quranapp.android.utils.Log
import com.quranapp.android.utils.extensions.color
import com.quranapp.android.utils.extensions.dp2px
import com.quranapp.android.utils.extensions.removeView
import com.quranapp.android.utils.extensions.setTextColorResource
import com.quranapp.android.utils.extensions.setTextSizePx
import com.quranapp.android.utils.extensions.updatePaddings
import com.quranapp.android.views.homepage2.HomepageCollectionLayoutBase
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class ReadHistoryLayout @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : HomepageCollectionLayoutBase(context, attrs, defStyleAttr), Destroyable {
    private var dbHelper: ReadHistoryDBHelper? = null
    private var adapter: ADPReadHistory? = null
    private var firstTime = true
    override fun destroy() {
        dbHelper?.close()
    }

    override fun getHeaderTitle(): Int {
        return R.string.strTitleReadHistory;
    }

    override fun getHeaderIcon(): Int {
        return R.drawable.dr_icon_history;
    }

    override fun initialize() {
        super.initialize()

        dbHelper = ReadHistoryDBHelper(context)
    }

    override fun setupHeader(header: LytHomepageTitledItemTitleBinding) {
        header.titleIcon.setColorFilter(context.color(R.color.colorPrimary))
    }

    override fun onViewAllClick() {
        context.startActivity(Intent(context, ActivityReadHistory::class.java))
    }

    override fun refresh(quranMeta: QuranMeta) {
        if (firstTime) {
            showLoader()
            firstTime = false
        }

        CoroutineScope(Dispatchers.IO).launch {
            val histories = dbHelper?.getAllHistories(10) ?: return@launch

            withContext(Dispatchers.Main) {
                hideLoader()
                if (histories.isEmpty()) {
                    makeNoHistoryAlert()
                } else {
                    if (adapter == null) {
                        adapter = ADPReadHistory(context, quranMeta, histories, Dimen.dp2px(context, 280f))
                        resolveListView().adapter = adapter
                    } else {
                        adapter!!.updateModels(histories)
                    }
                }
            }
        }
    }

    private fun makeNoHistoryAlert() {
        findViewById<View>(R.id.list).removeView()
        adapter = null

        if (findViewById<View?>(R.id.text) != null) {
            return
        }

        addView(AppCompatTextView(context).apply {
            id = R.id.text
            updatePaddings(context.dp2px(20f), context.dp2px(15f))
            setTextSizePx(R.dimen.dmnCommonSize1_5)
            setTextColorResource(R.color.colorText2)
            setTypeface(Typeface.SANS_SERIF, Typeface.ITALIC)
            layoutParams = ViewGroup.LayoutParams(MATCH_PARENT, WRAP_CONTENT)
            setText(R.string.strMsgReadShowupHere)
        })
    }

    override fun resolveListView(): RecyclerView {
        findViewById<View>(R.id.text).removeView()
        return super.resolveListView()
    }

}